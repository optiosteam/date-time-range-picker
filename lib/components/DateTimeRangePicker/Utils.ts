import moment, {Moment} from 'moment'

class Utils {
    static getDateTime (hasDate: boolean, hasTime: boolean, date?: Moment, time?: Moment) {
        const newDateTime = moment()
        if (hasDate) {
            if (! date) {
                return
            }

            newDateTime.year(date.year()).month(date.month()).date(date.date())
        }

        if (hasTime) {
            if (! time) {
                return
            }

            newDateTime.hours(time.hours()).minutes(time.minutes())
        }

        return newDateTime
    }
}

export default Utils
