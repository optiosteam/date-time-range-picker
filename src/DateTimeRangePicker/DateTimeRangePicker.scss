$color-highlight: #428be3;
$color-background-day: rgba(black, .1);
$color-other-month: rgba(black, .4);
$color-disabled: rgba(black, .4);

.date-time-range-picker {
  position: relative;
  display: flex;

  button {
    display: block;
    background-color: transparent;
    border: 0;
    outline: none;
    font-size: 1em;
    padding: 0;
    margin: 0;
  }

  .date-time-range-picker-times {
    display: flex;

    > div + div {
      margin-left: 1em;
    }
  }

  .date-time-range-picker-dates + .date-time-range-picker-times {
    flex-direction: column;
    padding: 2.9em 0 0 1em;

    > div + div {
      margin-left: 0;
      margin-top: 1em;
    }
  }
}
