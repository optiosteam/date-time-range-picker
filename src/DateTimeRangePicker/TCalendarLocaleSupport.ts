export type TCalendarLocaleSupport = 'nl' | 'en' | 'de' | 'fr'
