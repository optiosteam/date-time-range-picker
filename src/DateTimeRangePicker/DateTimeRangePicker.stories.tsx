import React from 'react'
import DateTimeRangePicker from './DateTimeRangePicker'
import moment from 'moment'

export default {
    title: 'DateTimeRangePicker',
    component: DateTimeRangePicker
}
const someDate = moment('2021/06/15')

export const basic = () => <DateTimeRangePicker
    calendarLocale={'fr'}
    initialDate={someDate}
    onChange={(date, date2) => console.log(date, date2)}
    onMonthChange={(newDate) => {
        console.log('Month changed: ', newDate)
    }}
/>
export const dateTime = () => (
    <DateTimeRangePicker
        time
        onChange={(date, date2) => console.log(date, date2)}
        onMonthChange={(newDate) => {
    console.log('Month changed: ', newDate)
}}/>
)
export const range  = () => (
    <DateTimeRangePicker
        range={true}
        onChange={(date, date2) => console.log(date, date2)}
        onMonthChange={(newDate) => {
    console.log('Month changed: ', newDate)
}} />
)
export const rangeWithMultipleMonths  = () => (
    <DateTimeRangePicker
        range={true}
        months={3}
        onChange={(date, date2) => console.log(date, date2)}
        onMonthChange={(newDate) => {
        console.log('Month changed: ', newDate)
    }} />
    )
