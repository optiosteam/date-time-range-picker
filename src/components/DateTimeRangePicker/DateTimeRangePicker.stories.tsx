import React from 'react'
import DateTimeRangePicker from './DateTimeRangePicker';

export default {
    title: 'DateTimeRangePicker',
    component: DateTimeRangePicker
}

export const basic = () => <DateTimeRangePicker/>
export const range  = () => <DateTimeRangePicker range={true} />
