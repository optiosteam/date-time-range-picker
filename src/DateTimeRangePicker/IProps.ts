import {Moment} from 'moment'
import {TCalendarLocaleSupport} from './TCalendarLocaleSupport'
import {IMonth} from './DatePicker/interfaces/IMonth'
import {IData} from './DatePicker/interfaces/IData'

interface IProps {
    date?: boolean,
    data?: IData[],
    time?: boolean,
    range?: boolean,
    inline?: boolean,
    fromDate?: Moment,
    untilDate?: Moment,
    initialDate?: Moment
    months?: number,
    onChange?: (fromDateTime: Moment, untilDateTime?: Moment) => void
    onMonthChange?: (month: { month: string; convertedDate: string; momentObject: moment.Moment }[]) => void
    calendarLocale?: TCalendarLocaleSupport
}

export default IProps
