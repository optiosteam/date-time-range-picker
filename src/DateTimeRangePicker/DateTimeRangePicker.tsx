import React, {useEffect} from 'react'
import moment, {Moment} from 'moment'

import './DateTimeRangePicker.scss'
import IProps from './IProps'
import TimePicker from './TimePicker/TimePicker'
import DatePicker from './DatePicker/DatePicker'
import Utils from './Utils'

const DateTimeRangePicker: React.FunctionComponent<IProps> = (
    {
        date = true,
        range = false,
        time = false,
        inline = true,
        fromDate,
        untilDate,
        months = 1,
        onChange
    }
) => {
    const [currentFromDate, setCurrentFromDate] = React.useState<Moment>(fromDate ? fromDate.clone() : undefined)
    const [currentUntilDate, setCurrentUntilDate] = React.useState<Moment>(untilDate ? untilDate.clone() : undefined)
    const [currentFromTime, setCurrentFromTime] = React.useState<Moment>(fromDate ? fromDate.clone() : undefined)
    const [currentUntilTime, setCurrentUntilTime] = React.useState<Moment>(untilDate ? untilDate.clone() : undefined)

    useEffect(
        () => {
            if (! onChange) {
                return
            }

            const newFromDate = Utils.getDateTime(date, time, currentFromDate, currentFromTime)
            if (! newFromDate) {
                return
            }

            if (! range) {
                onChange(newFromDate)

                return
            }

            const newUntilDate = Utils.getDateTime(date, time, currentUntilDate, currentUntilTime)
            if (! newUntilDate) {
                return
            }

            onChange(newFromDate, newUntilDate)
        }
    )

    return <div className={'date-time-range-picker'}>
        {date
            ? <DatePicker
                range={range}
                months={months}
                fromDate={currentFromDate}
                untilDate={currentUntilDate}
                onFromDateChanged={
                    (changedDate: Moment) => {
                        setCurrentFromDate(changedDate)
                    }
                }
                onUntilDateChanged={
                    (changedDate: Moment) => {
                        setCurrentUntilDate(changedDate)
                    }
                }
            />
            : null
        }

        {
            time
                ? <div className={'date-time-range-picker-times'}>
                    <div>
                        <label>{range ? 'From' : 'Time'}</label>
                        <TimePicker
                            step={15}
                            time={currentFromTime}
                            onTimeChanged={
                                (changedTime: Moment) => {
                                    setCurrentFromTime(changedTime)
                                    if (! range) {
                                        setCurrentUntilTime(changedTime)
                                    }
                                }
                            }
                        />
                    </div>
                    {
                        range
                            ? <div>
                                <label>To</label>
                                <TimePicker
                                    step={15}
                                    time={currentUntilTime}
                                    onTimeChanged={
                                        (changedTime: Moment) => {
                                            setCurrentUntilTime(changedTime)
                                        }
                                    }
                                />
                            </div>
                            : null
                    }
                </div>
                : null
        }
    </div>
}

export default DateTimeRangePicker
