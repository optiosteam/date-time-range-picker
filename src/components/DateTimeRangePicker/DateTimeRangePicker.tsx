import React, {useEffect, useState} from 'react'
import moment, {Moment} from 'moment'

import './DateTimeRangePicker.scss'
import IProps from './IProps'
import TimePicker from './TimePicker/TimePicker'
import DatePicker from './DatePicker/DatePicker'
import Utils from './Utils'

const DateTimeRangePicker: React.FunctionComponent<IProps> = (
    {
        date = true,
        range = false,
        time = false,
        inline = true,
        fromDate,
        untilDate,
        months = 1,
        onChange
    }
) => {
    const [currentFromDate, setCurrentFromDate] = useState<Moment | undefined>(fromDate ? fromDate.clone() : undefined)
    const [currentUntilDate, setCurrentUntilDate] = useState<Moment | undefined>(untilDate ? untilDate.clone() : undefined)
    const [currentFromTime, setCurrentFromTime] = useState<Moment | undefined>(fromDate ? fromDate.clone() : undefined)
    const [currentUntilTime, setCurrentUntilTime] = useState<Moment | undefined>(untilDate ? untilDate.clone() : undefined)
    const [isMounted, setIsMounted] = useState<boolean>(false)

    useEffect(
        () => {
            if (! isMounted) {
                setIsMounted(true)

                return
            }

            if (! onChange) {
                return
            }

            const newFromDate = Utils.getDateTime(date, time, currentFromDate, currentFromTime)
            if (! newFromDate) {
                return
            }

            
            const newUntilDate = Utils.getDateTime(date, time, currentUntilDate, currentUntilTime)
            if (! newUntilDate) {
                onChange(newFromDate, newFromDate)
                return
            }
            onChange(newFromDate, newUntilDate)
        },
        [currentFromDate, currentFromTime, currentUntilDate, currentUntilTime]
    )

    return <div className={'date-time-range-picker'}>
        {date
            ? <DatePicker
                range={range}
                months={months}
                fromDate={currentFromDate}
                untilDate={currentUntilDate}
                onFromDateChanged={
                    (changedDate: Moment | undefined) => {
                        setCurrentFromDate(changedDate)
                    }
                }
                onUntilDateChanged={
                    (changedDate: Moment | undefined) => {
                        setCurrentUntilDate(changedDate)
                    }
                }
            />
            : null
        }

        {
            time
                ? <div className={'date-time-range-picker-times'}>
                    <div>
                        <label>{range ? 'From' : 'Time'}</label>
                        <TimePicker
                            step={15}
                            time={currentFromTime}
                            onTimeChanged={
                                (changedTime: Moment) => {
                                    setCurrentFromTime(changedTime)
                                    if (! range) {
                                        setCurrentUntilTime(changedTime)
                                    }
                                }
                            }
                        />
                    </div>
                    {
                        range
                            ? <div>
                                <label>To</label>
                                <TimePicker
                                    step={15}
                                    time={currentUntilTime}
                                    onTimeChanged={
                                        (changedTime: Moment) => {
                                            setCurrentUntilTime(changedTime)
                                        }
                                    }
                                />
                            </div>
                            : null
                    }
                </div>
                : null
        }
    </div>
}

export default DateTimeRangePicker
