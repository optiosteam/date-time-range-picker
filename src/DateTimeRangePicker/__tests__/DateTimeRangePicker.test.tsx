import React from 'react'
import { shallow } from 'enzyme'

import DateTimeRangePicker from '../DateTimeRangePicker'

describe('DateTimeRangePicker', () => {
    it('should render correctly', () => {
        const component = shallow(<DateTimeRangePicker theme={'primary'} />)

        expect(component).toMatchSnapshot()
    })
})
