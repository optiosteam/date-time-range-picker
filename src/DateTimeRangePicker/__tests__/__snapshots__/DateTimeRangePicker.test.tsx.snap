// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeRangePicker should render correctly 1`] = `
<div
  className="date-time-range-picker"
>
  <DatePicker
    fromDate={"2019-10-15T16:23:15.000Z"}
    months={2}
    onFromDateChanged={[Function]}
    onUntilDateChanged={[Function]}
    range={false}
    untilDate={"2019-10-15T16:23:15.000Z"}
  />
</div>
`;
